Expose Solution Operator config loading as a standalone Go module

gdemonet opened the pull request:
**Component**: go, solutions

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:

Solution Operators are expected to read a config file on startup which lists the available image repositories for the Solution components. This config file format belongs to the Solution framework, and as such, it makes sense for said framework to expose an _importable_ package for the Operators to use.

**Summary**:

- Define a new Go module `github.com/scality/metalk8s/go/solution-operator`, including (for now) a single `config` package. Two methods are exposed for convenience:
	- `LoadConfiguration` which takes a `Reader` interface and returns a parsed and validated `OperatorConfig`,
	- and `LoadConfigurationFromFile` which takes a file path and passes its contents to `LoadConfiguration`
- Use this package in the `examples/metalk8s-solution-example/operator` example Solution Operator

**Acceptance criteria**: 

- The example Operator still works
- One can `go get` this new module and use the `config` package as expected

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2837 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

gdemonet committed:
go: Add module for Solution Operators to use

Only exposing a `config` package for now, the module is simply named
`github.com/scality/metalk8s/go/solution-operator`.

See: #2837

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2837-go-mod-operator-config` into
`development/2.6`
* [w/2.7/improvement/2837-go-mod-operator-config](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2837-go-mod-operator-config) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

miniscruff commented on go/solution-operator/config/config_test.go:
```diff
@@ -0,0 +1,174 @@
+		"invalid-no-apiversion": {
+			filename: "invalid-no-apiversion.yaml",
+			result:   nil,
+			error:    "missing required field: apiVersion",
```
Still new to go but the way we have been testing errors in zenko-operator has been nice. Instead of using the string of the error, we have been creating error values and using github.com/pkg/errors wrap method to add extra values. Then errors.cause to check the root error. Avoids having to copy paste strings and allows you to change error messages without having to also change tests just for a string value.

NicolasT replied:
For reference, https://blog.golang.org/go1.13-errors

gdemonet replied:
Thanks for the suggestion, I'll have a look!

gdemonet replied:
So I'll go with the "plain" Go 1.13 `Unwraps` as suggested in Nicolas' link, which should achieve the same effect as `github.com/pkg/errors` without requiring any additional package.

gdemonet replied:
@miniscruff I explained a bit more the choices around error handling in [`9b6d7b5` (#2844)](https://github.com/scality/metalk8s/pull/2844/commits/9b6d7b5e834cc540c6765d5d387f46f9cdec6360), which doesn't really match what is done for Zenko Operator.
In any case, thanks for pointing out that improvements could be made, I learnt a great deal while playing with all this!

miniscruff reviewed (dismissed):
Looking good to me, I might have an extra comment or two once I get to use it but as far as I can tell the API is simple and straightforward.

NicolasT commented:
Since `solution-operator` is not really an operator, did you consider `go/solution-operator-lib` instead as the package root? Could then also use `go/solution-operator-lib/pkg/config` as the package, so if there's a need for `go/solution-operator-lib/internal/...` at some point in time this can be trivially added.

gdemonet commented:
> Since `solution-operator` is not really an operator, did you consider `go/solution-operator-lib` instead as the package root? Could then also use `go/solution-operator-lib/pkg/config` as the package, so if there's a need for `go/solution-operator-lib/internal/...` at some point in time this can be trivially added.

Good points, will change these.

slaperche-scality commented on examples/metalk8s-solution-example/operator/go.sum:
```diff
@@ -42,13 +42,15 @@ github.com/Azure/go-autorest/autorest/date v0.2.0 h1:yW+Zlqf26583pE43KhfnhFcdmSW
 github.com/Azure/go-autorest/autorest/date v0.2.0/go.mod h1:vcORJHLJEh643/Ioh9+vPmf1Ij9AEBM5FuBIXLmIy0g=
 github.com/Azure/go-autorest/autorest/mocks v0.1.0/go.mod h1:OTyCOPRA2IgIlWxVYxBee2F5Gr4kF2zd2J5cFRaIDN0=
 github.com/Azure/go-autorest/autorest/mocks v0.2.0/go.mod h1:OTyCOPRA2IgIlWxVYxBee2F5Gr4kF2zd2J5cFRaIDN0=
+github.com/Azure/go-autorest/autorest/mocks v0.3.0 h1:qJumjCaCudz+OcqE9/XtEPfvtOjOmKaui4EOpFI6zZc=
```
Not sure why we have all those changes to the `go.sum` of `solution-example` (only additions, no removal).

gdemonet replied:
I don't understand how `go.sum` works, TBH... I've removed the code from `pkg/config` and ran `go mod tidy` (I expected this would refresh both `go.mod` and `go.sum` based on the imports within the module), which complains about something entirely unrelated:

```
example-solution-operator imports
        sigs.k8s.io/controller-tools/pkg/crd/generator: module sigs.k8s.io/controller-tools@latest found (v0.4.0), but does not contain package sigs.k8s.io/controller-tools/pkg/crd/generator
```

And that's roughly all I did, so I'm not sure what's expected or what I should do.

gdemonet replied:
OK I split the `go mod tidy` into its own commit, to make things clearer.

bert-e commented:
# History mismatch



Merge commit #5d26f997b79b6931cd726bf7e2cae92bb771863d on the integration branch
`w/2.7/improvement/2837-go-mod-operator-config` is merging a branch which is neither the current
branch `improvement/2837-go-mod-operator-config` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2837-go-mod-operator-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
go/operator-lib/config: Rework error handling

Following advice from https://blog.golang.org/go1.13-errors, we make
sure that errors raised from this package are easy to handle:
- they either wrap an error the caller can manage (e.g.
  `os.ErrNotExist`) or a publicly exposed `ErrValidation`
- the validation error messages are consistent and always include the
  key(s) which caused the validation error

Note that we don't want to explicitly wrap the internal errors (like
`errMissingField`) because they are not public, hence we don't expect
this API to be stable and don't want to test against them.
On the other hand, the user-facing messages are part of the public API,
which is why we still test them.

See: #2837

miniscruff reviewed (dismissed):
Looks good to me, and worked when I tried manually.

gdemonet commented:
/reset

gdemonet commented:
/approve

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2837-go-mod-operator-config` into
`development/2.6`
* [w/2.7/improvement/2837-go-mod-operator-config](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2837-go-mod-operator-config) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/4a0d7fa4b9bd6b6c8c8bcb6b9d2697a64b60a90c) did not succeed in branch w/2.7/improvement/2837-go-mod-operator-config.




*The following options are set:* **approve**

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2837-go-mod-operator-config` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/2837-go-mod-operator-config`
* Rebase `improvement/2837-go-mod-operator-config` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #a883fa2810b334734175867fe82f82ede5508e42 on the integration branch
`w/2.7/improvement/2837-go-mod-operator-config` is merging a branch which is neither the current
branch `improvement/2837-go-mod-operator-config` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2837-go-mod-operator-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2837-go-mod-operator-config` into
`development/2.6`
* [w/2.7/improvement/2837-go-mod-operator-config](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2837-go-mod-operator-config) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal










*The following options are set:* **approve**

miniscruff reviewed (dismissed):
Code looks to be unchanged with the conflicts so still good to me, can run another manual test if you think I should.

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/d5995e9e4abf3b0f847a0e27cd5ee1067e5eeed4) did not succeed in branch w/2.7/improvement/2837-go-mod-operator-config.




*The following options are set:* **approve**

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2837-go-mod-operator-config` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/2837-go-mod-operator-config`
* Rebase `improvement/2837-go-mod-operator-config` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #5590a4ae43ae77d9ff4b8cdb421a93182848c12b on the integration branch
`w/2.7/improvement/2837-go-mod-operator-config` is merging a branch which is neither the current
branch `improvement/2837-go-mod-operator-config` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2837-go-mod-operator-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet committed:
example-solution/operator: Update config loading

We now use the public Go module
`github.com/scality/metalk8s/go/solution-operator-lib`, from which we
use the `pkg/config` package to replace the one that was previously
defined in `pkg/config` of this Operator code.

See: #2837

gdemonet committed:
example-solution/operator: Run `go mod tidy`

See: #2837

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2837-go-mod-operator-config` into
`development/2.6`
* [w/2.7/improvement/2837-go-mod-operator-config](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2837-go-mod-operator-config) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye gdemonet.
